app/database/ancestors.go

```diff
@@ -0,0 +1,178 @@
+	markAsProcessing, err := s.db.CommonDB().Prepare(query)
+	mustNotBeError(err)
+	defer func() { mustNotBeError(markAsProcessing.Close()) }()
+
+	bIsSelfColumn := ""
+	bIsSelfValue := ""
+	if objectName == "groups" {
+		bIsSelfColumn = ", bIsSelf"
+		bIsSelfValue = ", '0' AS bIsSelf"
+	}
+
+	// For every object marked as 'processing', we compute all its ancestors
+	query = `
+		INSERT IGNORE INTO ` + objectName + `_ancestors
+		(
+			id` + upObjectName + `Ancestor,
+			id` + upObjectName + `Child` + `
+			` + bIsSelfColumn + `
+		)
+		SELECT
+			` + relationsTable + `.id` + upObjectName + `Parent,
+			` + relationsTable + `.id` + upObjectName + `Child
+			` + bIsSelfValue + `
+		FROM ` + relationsTable + `
+		JOIN ` + objectName + `_propagate
+		ON (
+			` + relationsTable + `.id` + upObjectName + `Child = ` + objectName + `_propagate.ID OR
+			` + relationsTable + `.id` + upObjectName + `Parent = ` + objectName + `_propagate.ID
+		)
+		WHERE
+			` + objectName + `_propagate.sAncestorsComputationState = 'processing'` + groupsAcceptedCondition + `
+		UNION ALL
+		SELECT
+			` + objectName + `_ancestors.id` + upObjectName + `Ancestor,
+			` + relationsTable + `_join.id` + upObjectName + `Child
+			` + bIsSelfValue + `
+		FROM ` + objectName + `_ancestors
+		JOIN ` + relationsTable + ` AS ` + relationsTable + `_join ON (
+			` + relationsTable + `_join.id` + upObjectName + `Parent = ` + objectName + `_ancestors.id` + upObjectName + `Child
+		)
+		JOIN ` + objectName + `_propagate ON (
+			` + relationsTable + `_join.id` + upObjectName + `Child = ` + objectName + `_propagate.ID
+		)
+		WHERE 
+			` + objectName + `_propagate.sAncestorsComputationState = 'processing'`
+	if objectName == "groups" {
+		query += `
+			AND (
+				groups_groups_join.sType = 'invitationAccepted' OR
+				groups_groups_join.sType = 'requestAccepted' OR
+				groups_groups_join.sType = 'direct'
+			)
+		UNION ALL
+		SELECT
+			` + `groups_propagate.ID AS idGroupAncestor,
+			` + `groups_propagate.ID AS idGroupChild,
+			'1' AS bIsSelf
+		FROM groups_propagate
+		WHERE groups_propagate.sAncestorsComputationState = 'processing'`
+	}
+
+	insertAncestors, err := s.db.CommonDB().Prepare(query)
+	mustNotBeError(err)
+	defer func() { mustNotBeError(insertAncestors.Close()) }()
+
+	// Objects marked as 'processing' are now marked as 'done'
+	query = `
+		UPDATE ` + objectName + `_propagate
+		SET sAncestorsComputationState = 'done'
+		WHERE sAncestorsComputationState = 'processing'`
+	markAsDone, err := s.db.CommonDB().Prepare(query)
+	mustNotBeError(err)
+	defer func() { mustNotBeError(markAsDone.Close()) }()
+
+	for hasChanges {
+		_, err = markAsProcessing.Exec()
+		mustNotBeError(err)
+		_, err = insertAncestors.Exec()
+		mustNotBeError(err)
+
+		var result sql.Result
+		result, err = markAsDone.Exec()
+		mustNotBeError(err)
+		var rowsAffected int64
+		rowsAffected, err = result.RowsAffected()
+		mustNotBeError(err)
+		hasChanges = rowsAffected > 0
+	}
```

Comment:
Don't you think it may be useful to put a guard in such loops to make sure it stops even if something goes wrong? (i.e. stop anyway after 50 iterations and log an error if so)

Reply:
discussed this in Slack